### ISSUE TITLE
Default interface methods

### ISSUE THREAD

**Issue by cartermp** [fsharp/fslang-suggestions#679]
I propose we add support for consumption and production of methods on interfaces with concrete implementations to enable the following:

* Interoperation with C# and .NET when [it is implemented](https://github.com/dotnet/csharplang/issues/52)
* Help API authors version interfaces
* Aid interoperation with Java and iOS for Xamarin programming scenarios
* Act as a basis for trait-like programming, potentially using this as a vehicle for further abstractions

The existing way of approaching this problem in F# is to either:

* Use abstract classes; though realistically, this is not done much because it forces implementation inheritance.
* Not communicate via interfaces.

Additionally, it is important to call out that this **requires a runtime change for .NET**. The change for the runtime is currently being driven and coordinated by the C# and .NET runtime teams.

## Description

The syntax for interfaces would be extended to permit the following:

* A concrete body for a member (i.e., a default implementation)
* Explicit access modifiers
* Overriding members
* Static members

❗️ The following syntax and concepts are a **starting point**, and absolutely need discussion. ❗️ 

### Concrete member bodies

To begin, there needs to be a way to specify a concrete body for a member in an interface:

```fsharp
type IA =
    default M() = printfn "IA.M()"
```

A class that implements this interface need not implement its concrete method:

```fsharp
type C() =
    interface IA

let ia: IA = C()
ia.M() // Prints "IA.M()"
```

The same would apply for object expressions:

```fsharp
type IA =
    default M() = printfn "IA.M()"

type IB =
    inherit IA
    abstract N: unit -> unit

let x =
    { new IB with
          member __.N() = () }

x.M() // Prints "IA.M()"
```

Note that a class does not inherit members from the interface, so the following code would be illegal:
```
C().M() // ERROR, 'C' does not contain member 'M'
```

### Member modifiers

The default modifier for interface members is `public`. That is, the following member is public:

```fsharp
type IA =
    default M() = printfn "IA.M()"
```

The following access modifiers could be added: `public`, `internal`, and `private`:

```fsharp
type IModifiers =
    abstract internal Doop: unit
    abstract private Hoopty: unit
```

This is keeping in line with access modifiers being added in C#.

### Overriding members

Interfaces could override default members from other interfaces:

```fsharp
type IA
    default M() = printfn "IA.M()"

type IB
    inherit IA
    override IA.M() =  printfn "IB.M()" // explicitly named

type IC
    inherit IA
    override M() = printfn "IA.M()" // implicitly named
```

**Open question:** Should explicit naming like this even be a thing?

Overrides are useful to provide more useful "versions" of methods. For example, a new `First()` method on `IEnumerable` may have a much more efficient implementation on the interface `IList`.

Note that no overriding would occur unless specified with the keyword `override`.

### Reabstraction

A default member in an interface could also be overridden to be made abstract again:

```fsharp
type IA =
    default M() = printfn "IA.M()"

type IB =
    inherit IA
    override abstract M: unit -> unit

type C() =
    inherit IB // Error: 'C' does not implement member 'IA.M'
```
If permitted, the `abstract` keyword should be required. This is currently also an [open issue for C#](https://github.com/dotnet/csharplang/blob/master/proposals/default-interface-methods.md#reabstraction). We should only implement this if C# does.

### Most specific override

I would imagine we do the same as C# here:

> We require that every interface and class have a _most specific override_ for every virtual member among the overrides appearing in the type or its direct and indirect interfaces. The _most specific override_ is a unique override that is more specific than every other override. If there is no override, the member itself is considered the most specific override.

For example:

```fsharp
type IA
    default M() = printfn "IA.M()"

type IB
   inherit IA
   override IA.M() = printfn "IB.M()"

type IC
    inherit IA
    override IA.M() = printfn "IC.M()"

interface ID
    inherit IB
    inherit IC  // ERROR: no most specific override for 'IA.M'

type C()
    inherit IB
    inherit IC // ERROR: no most specific override for 'IA.M'

type D() // OK
    inherit IA
    inherit IB
    abstract M: unit -> unit
```

This rule ensures that ambiguities from diamond inheritance are resolved by the programmer.

### `static`  and `private` members

Because interfaces could have concrete default implementations, common code could be abstracted into `static` and `private` members. There are some [open issues](https://github.com/dotnet/csharplang/blob/master/proposals/default-interface-methods.md#static-and-private-methods) for C# here today.

This could, in effect, turn interfaces into something in between what we use today and abstract classes. It's a bit iffy to me. I wouldn't want something like `let`-bound values in them, either.

### Base interface invocation

Code in a type that derives from an interface with a default method can explicitly invoke that interface's "base" implementation:

```fsharp
type IA
    default M() = printfn "IA.M()"

type IB
    inherit IA
    override M() = printfn "IB.M()"

type IC
    inherit IA
    override M() = printfn "IC.M()"

type ID
    inherit IB
    inherit IC
    override IA.M() = IB.base.M()
```

Note that the syntax is `Type.base.Member`. This is, in keeping with the theme of this suggestion, in line with the [C# spec for the feature](https://github.com/dotnet/csharplang/blob/master/proposals/default-interface-methods.md#base-interface-invocations).

Additionally, the C# team has verified that these changes would **not** affect existing programs. I would expect anything we do for F# also not affect existing programs.

## Pros and Cons 

The advantages of making this adjustment to F# are:

* Ability to version interfaces over time
* Ability to begin writing trait-like programs
* Nicer interop with Android and iOS for Xamarin
* Interoperation with the direction C# and the CLR would take
* A possible vehicle for higher abstractions

The disadvantages of making this adjustment to F# are

* Changes to a fundamental type (interfaces) put cognitive load on people that is arguably unnecessary
* This feels like it solves less of a problem for F# than it does for C#
* It's considerable work to implement, get right, and document

## Extra information

Estimated cost (XS, S, M, L, XL, XXL): XL

Related information:

* C# proposal champion issue: https://github.com/dotnet/csharplang/issues/52
* C# spec for [default interface methods](https://github.com/dotnet/csharplang/blob/master/proposals/default-interface-methods.md)

## Affidavit (please submit!)

Please tick this by placing a cross in the box:
* [x] This is not a question (e.g. like one you might ask on [stackoverflow](http://stackoverflow.com)) and I have searched stackoverflow for discussions of this issue
* [x] I have [searched both open and closed suggestions on this site](http://github.com/fsharp/fslang-suggestions/issues) and believe this is not a duplicate
* [x] This is not something which has obviously "already been decided" in previous versions of F#.  If you're questioning a fundamental design decision that has obviously already been taken (e.g. "Make F# untyped") then please don't submit it.

Please tick all that apply:
* [x] This is not a breaking change to the F# language design
* [x] I or my company would be willing to help implement and/or test this

**Comment by dsyme** [comment]
The listed benefits are these:

1. Help API authors version interfaces
2. Aid interoperation with Java and iOS for Xamarin programming scenarios
3. Act as a basis for trait-like programming, potentially using this as a vehicle for further abstractions

Of these, only (2) is of interest for me, i.e. this feature may be necessary for interoperating with .NET code.  This require consumption only.  

I simply don't believe (1) and (3) are real benefits in the overall scheme of things vis.a.vis the massive  complexifying costs of software designs based around implementation inheritance, especially in a language like F# that seeks to re-orient programmers minds away from this horror.  An over-focus on implementation inheritance has cost the software industry billions and billions of dollars in lost productivity and frankly is has created a lost generation of programmers who obsessively make complex things out of complex things.   .NET made the right decision in version 1 in not allowing interfaces to have implementations, and to be honest I think it should stick with it, placing a feature like this at the outer fringes of the .NET universe like COM interop, dynamic and that whacky stuff the .NET team did with interfaces for Office interop in C# 5 which everyone has forgotten.

Additionally, the feature is *massively* controversial in the C# world, with https://github.com/dotnet/csharplang/issues/288 having the highest proportion of downvotes I've seen for a C# language design proposal that's being taken forward.

So, in short, this feature would be interop-only, placing it in the same class as "protected" (which again encourages implementation inheritance).

**Comment by cartermp** [comment]
@dsyme Just to say (not sure if this is clear), I wanted to capture all possibilities with this feature in the issue. My primary concern is interoperability, but I wanted to "open the floor" to what others thing about this as more than just as an interop feature.

**Comment by dsyme** [comment]
@cartermp yes totally understood, we have to track the feature for interop purposes, and I know you're not championing it in full as such :) Thanks for doing it and adding the detail :)

**Comment by dsyme** [comment]
@Horusiath thanks for that info. The C# folk should really be looking very closely at the overall Scala experience with this feature.

**Comment by tpetricek** [comment]
Just to derail the discussion a bit, I would really love to be able to implement interfaces by saying `interface ISomething = <expr>` where `<expr>` is an expression returning the interface implementation.

I think that might actually cover some of the use cases of the C# proposal, because you could do:

```
type IA = 
  abstract M : unit -> unit
  abstract N : unit -> unit

let makeIA f = 
  { new IA with 
      member x.M() = f ()
      member x.N() = printfn "default" }

type C() = 
  interface IA = makeIA (fun () -> printfn "C")
```

And you could do this without introducing any more complex object-oriented concepts...

**Comment by dsyme** [comment]
@tpetricek I really like that suggestion - there is a separate suggestion for it somewhere.  TBH it is vastly more "in tune" with F# methodology (delegation over inheritance) than the default interfaces feature.

**Comment by cartermp** [comment]
Agreed, that does feel much more natural than the override/default goop that would come with a direct port of the C# feature set.

**Comment by wallymathieu** [comment]
If you have the ability to specify static for interface members (in c# and f#), you could perhaps add module interfaces.

**Comment by zpodlovics** [comment]
Probably also worth exploring and discussing this (here or as a new issue) C# lang issue to have a proper concepts and mappings in F# (it also explores interface based generic numeric code as an example usage similar to @tpetricek http://tomasp.net/blog/fsharp-generic-numeric.aspx/):

"Roles, extension interfaces and static interface members

This is an attempt to address the scenarios targeted by my previous "shapes" investigation (which was in turn inspired by the "Concept C#" work by Claudio Russo and Matt Windsor), but in a way that leverages interfaces rather than a new abstraction mechanism. It is not necessary to read the previous proposals in order to understand this one."

https://github.com/dotnet/csharplang/issues/1711

**Comment by robkuz** [comment]
@cartermp did I read this correctly: Default Interfaces wont be available on the Desktop? Why is this?

**Comment by SimplerSoftware** [comment]
@robkuz - I guess one thing to consider is that desktop support is coming to DotNet core 3.0, and so perhaps it will be less of an issue since you could just target DotNet core at that point: [https://blogs.msdn.microsoft.com/dotnet/2018/05/07/net-core-3-and-support-for-windows-desktop-applications/](https://blogs.msdn.microsoft.com/dotnet/2018/05/07/net-core-3-and-support-for-windows-desktop-applications/)

**Comment by jnm2** [comment]
You'd never be able to use Default Interface Methods in a library that targets .NET Standard, though, because it would blow up at runtime on .NET Framework and UWP. In order to use DIM, a library wouldbe forced to single-target .NET Core. (As a library author, that's a hard sell.)

**Comment by cartermp** [comment]
@robkuz Default interface methods require a runtime change. This means that there is also a check for seeing if the feature is supported by a given runtime: https://github.com/dotnet/csharplang/blob/master/proposals/default-interface-methods.md#clr-support-api

No shipped .NET Framework version supports this today, and it's highly unlikely that they ever will due to the risk of breaking existing apps that are so widespread. .NET Core will eventually have this in its runtime, but it's not completely resolved if it will also be in some future .NET Framework, mono, or UWP runtime. And as @jnm2 mentioned, unless every runtime that supports a .NET Standard also has this feature, then you wouldn't be able to use them in .NET Standard. It is not in the upcoming [.NET Standard 2.1](https://github.com/dotnet/standard/blob/master/docs/planning/netstandard-2.1/README.md) plan either.

The question in my mind, from a long-term planning perspective, is what we do beyond simply ensuring that we don't blow up in the face of such a construct. Is the feature a copy of C#? Probably not. A fully-fledged traits/typeclasses system? That would need a proper design that would take time. How would it be rationalized with existing things like SRTP? How to think about interfaces today vs. interfaces tomorrow vs. functions as interfaces vs. normal generics vs. SRTP vs. {insert thing here}? But at least in my mind, the mechanism for implementing _something_ is coming, so it would be good to think about what that _something_ is at a high level, what sort of behaviors it could have, and how to rationalize it with existing features in this space.

**Comment by zpodlovics** [comment]
Before this feature improvement / high level design planning starts, I would like to suggest to add the indirect calls supports to this list. Probably also worth to create an independent proposal/design document to the indirect calls for the details.

"We use ldftn + calli in lieu of delegates (which incur an object allocation) in performance-critical pieces of our code where there is a need to call a managed method indirectly. This change allowed method bodies with a calli instruction to be eligible for inlining. Our dependency injection framework generates such methods." [1] [2] [3]

"This proposal provides language constructs that expose low level IL opcodes that cannot currently be accessed efficiently, or at all: ldftn, ldvirtftn, ldtoken and calli. These low level op codes can be important in high performance code and developers need an efficient way to access them." [2] [3]

The high level design is fine as long as if it's done "without regret" (indirection / abstraction overhead). 

"It has been said that all problems in computer science can be solved by adding another level of indirection, except for performance problems, which are solved by removing levels of indirection."

**Compilers should be our tools for removing levels of indirection automatically.**

[1] https://blogs.msdn.microsoft.com/dotnet/2018/08/20/bing-com-runs-on-net-core-2-1/
[2] dotnet/coreclr#13756
[3] https://github.com/dotnet/csharplang/blob/master/proposals/intrinsics.md
[4] https://github.com/dotnet/csharplang/blob/master/proposals/intrinsics.md
[5] https://github.com/dotnet/csharplang/blob/3b43266cea6012b8964cb6ebc0e913ae7d6abd12/meetings/2017/LDM-2017-01-11.md
[6] https://github.com/dotnet/csharplang/blob/master/meetings/2018/LDM-2018-09-05.md

**Comment by HaloFour** [comment]
@zpodlovics 

#191

**Comment by zpodlovics** [comment]
@HaloFour You probably mean this: https://github.com/dotnet/csharplang/issues/191

**Comment by HaloFour** [comment]
@zpodlovics 

Oh yes, sorry, I mistakenly thought your comment was posted to that repo.

In that case I might suggest you open a new issue specifically to look into that same support for F#.

**Comment by sgf** [comment]
this is a great function. i like it.
Because the class itself unable to cross multiple inheritance
but interface can be.
in fact,<Default interface methods> is a Best implementation of <Favor Object composition over inheritance>.
Meanwhile, the code is the same,u dont need write too much code duplication.

**Comment by MichalStrehovsky** [comment]
> You'd never be able to use Default Interface Methods in a library that targets .NET Standard, though, because it would blow up at runtime on .NET Framework and UWP. In order to use DIM, a library would be forced to single-target .NET Core. (As a library author, that's a hard sell.)

Default interface methods are now being added to the .NET Standard: dotnet/standard#1019 :/

**Comment by abelbraaksma** [comment]
The original post here says that reabstraction is an open question and should only be implemented if C# does. 

@cartermp, it has meanwhile been answered and the C# team considers it a requirement to allow reabstraction. See the same link, section on Resolved Questions. Perhaps we can update the OP?

**Comment by cartermp** [comment]
@abelbraaksma We'll have to go over this one in more detail if it's to be a thing in .NET Standard 2.1. However, reabstraction isn't essential to interoperation. So far, my stance is that we need to be able to consume DIMs without blowing up. Implementing the same set of features is different, IMO.

**Comment by robkuz** [comment]
@cartermp 
I would like to understand how the stance on this feature synchronizes with seamless 2-way integration which is so often espoused as one of the highest goods/goals in F#.
What are the (objective) reasons for not implementing this feature? 
I must admit I find this troubling.

**Comment by cartermp** [comment]
What do you fine troubling?

**Comment by robkuz** [comment]
@cartermp many interesting have been marked as "awaiting C#" or "CLR adjustment needed" and therefore been rejected for now (?).

So all hope for any significant language improvements are being pushed to C#/CLR.
However if it is possible to knockout the "seamless-2-way-integration" mantra and deny the (producing) implementation of Default Interfaces because of subjective preferences then the same could easily be done for GADTs, HKTs and TCs (or anything that might come in the future).

I find this troubling and the message it sends is inconsistent to say the least.

**Comment by cartermp** [comment]
> So all hope for any significant language improvements are being pushed to C#/CLR.

This is not true. It may be true for those you care about, but certainly not for others.

**Comment by robkuz** [comment]
Actually those 3 I specifically named are among the top 6 most upvoted and also oldest. 
So it seems that this isn't exactly only me.

But then again your reply isn't answering my question. So again:

What are the **objective** criteria to ditch the 2-way-seamless integration of a C# feature?

This concerns this specific feature but also potentially other (future) features.

**Comment by cartermp** [comment]
There is little objective criteria in language design. That's why it's _design_, not _engineering_. Nonetheless, I will give my criteria, which likely differs a bit from @dsyme's:

* I don't think it's the right solution to problems around ad-hoc polymorphism on .NET today. If .NET goes down the route of ad-hoc polymorphism, Typeclasses via the [Shapes](https://github.com/dotnet/csharplang/issues/164) proposal solves more concrete problems that .NET programmers run into today. Adding DIMs and Shapes just complicates things needlessly.
* I think it's too object-oriented. Having a feature depend on inheritance-based features is a far cry from the pattern-based and function-oriented way people typically write F# code. This is similar to the lack of support for `protected` in F#, which is unlikely to be implemented due to it only being useful with inheritance-based hierarchies. Inheritance-based hierarchies are not something we wish to promote, and this feature also encroaches upon that.
* This lets people get a lot closer to multiple inheritance, which I don't really like.
* There is a belief in the C# community that interfaces are just fine the way they are, and that people can use abstract classes when they actually need what this feature would offer. I agree with that, and don't feel comfortable pushing something forward when so many people disagree with it.
* The motivations given for this feature - letting library authors version interfaces and easing Xamarin interop for Android - do not feel important enough to push this feature through given the controversy around it. Xamarin.

Lastly, the feature hasn't even shipped in a preview of C# yet. Given the controversy around it, I'd much rather wait until there is consensus that it's okay before we think of proceeding with anything other than not blowing up when we see it in metadata.

**Comment by Lanayx** [comment]
Wanted to leave comment here, I'm writing a [F# library](https://github.com/fsharplang-ru/pulsar-client-dotnet) which is used by both F# and C# developers. And I have to give users ability to provide their classes for routing, authentication, interceptors. Without support of DIM I'll be making breaking changes with new releases, so if I want to avoid it I need to extract this logic to C# code which is undesirable.

**Comment by cartermp** [comment]
Thanks for the feedback @Lanayx. I've come around to thinking that we should also just support producing DIMs, even though it's an inheritance-oriented feature. It feel useful enough (versioning interfaces) and interop with Java/iOS is valuable.

**Comment by cartermp** [comment]
Closing out as completed for F# 5. We may revisit creating these. The work to create them isn't terrible large, but we'd need to figure out a good design.

**Comment by realvictorprm** [comment]
I'm meanwhile in favour of supporting the creation of them in F# too.

**Comment by abelbraaksma** [comment]
Just summing up for visitors late to the party: currently consuming default interfaces is possible, creating them is not.
Considering the above comments, perhaps we need a new language suggestion for DIM creation from F#. 

PR (in F# 5.0): https://github.com/dotnet/fsharp/pull/8628
RFC FS-1074: https://github.com/fsharp/fslang-design/blob/master/FSharp-5.0/FS-1074-default-interface-member-consumption.md

**Comment by Serentty** [comment]
> Just summing up for visitors late to the party: currently consuming default interfaces is possible, creating them is not.
> Considering the above comments, perhaps we need a new language suggestion for DIM creation from F#.
> 
> PR (in F# 5.0): [dotnet/fsharp#8628](https://github.com/dotnet/fsharp/pull/8628)
> RFC FS-1074: https://github.com/fsharp/fslang-design/blob/master/FSharp-5.0/FS-1074-default-interface-member-consumption.md

I'd be happy to see such a proposal. I've come to really like this feature in Rust. Passing in lambdas really doesn't seem like the same thing at all, because it means allowing variation at the instance level, not just the type level. Plus it means carrying all that extra data around in the instance.
